### ISSUE TITLE
Support explicit registration of workers

### ISSUE THREAD

**Issue by ggoodman** [jasnell/piscina#106]
## Description

Add a mechanism for worker code to register itself with `piscina` instead of relying on the existing `./src/worker.ts` logic that attempts to resolve and load (and cache) worker code.

## Use-case

When bundling server-side code, `./dist/worker.js` will not be picked up in the same bundle as `./dist/index.js` because there are no statically-analyzable dependencies on the worker file. As a result, at runtime, the new location of the now-bundled `./dist/index.js` is no longer adjacent to `./dist/worker.js` on disk.

To get around this, it is possible to add `./dist/worker.js` as a separate entrypoint to make sure it gets tracked by the bundler. An alternative solution, seen in [workerpool](https://npm.im/workerpool) is that it exposes a [`worker(functions: Record<string, (...args: any[]) => any>): void`](https://github.com/josdejong/workerpool#worker) API that allows a worker to _import_ the library's coordination code. The way I see this is that it inverts the relationship so that the worker has an explicit opt-in option.

A side-benefit that I'd love to see considered in `piscina` is that this provides a clean facility for registering functions that are not necessarily default exports and even for registering different named functions so that a dispatching layer isn't needed.

**Comment by mcollina** [comment]
I think the use case should be supported.

(I usually recommend not to bundle server-side code, but anyway).

Would you like to send a PR?

**Comment by jasnell** [comment]
I'm still struggling a bit with what the API would be and how it would be different. Sorry, been a bit distracted the past few days so I may just be missing something.

Is the idea essentially that a worker could register itself by name as opposed to piscina using module resolution to acquire it? So, essentially something like:

```js
// worker.js
const Piscina = require('piscina');

module.exports = Piscina.worker('foo', async () => {
  // do stuff
});
```

Such that we can do something like...

```js
const Piscina = require('piscina');
require('./worker.js')

const pool = new Piscina();

pool.runTask({ a: 1, b: 2}, 'foo');
```

**Comment by jasnell** [comment]
I'm adding an example that demonstrates how this pattern can be achieved fairly easily without making changes to piscina itself... Specifically:

```js
// helper.js
function makeTask(op, ...args) {
  return { op, args };
}

function dispatcher(obj) {
  return async ({ op, args }) => {
    return await obj[op](...args);
  }
}

module.exports = {
  dispatcher,
  makeTask
};
```

```js
// worker.js
'use strict';

const { dispatcher } = require('./helper');

module.exports = dispatcher({
  add(a, b) { return a + b; },
  sub(a, b) { return a - b; }
});
```

```js
// index.js
'use strict';

const Piscina = require('../..');
const { resolve } = require('path');
const { makeTask } = require('./helper');

const piscina = new Piscina({
  filename: resolve(__dirname, 'worker.js')
});

(async function () {
  const result = await Promise.all([
    piscina.runTask(makeTask('add', 4, 6)),
    piscina.runTask(makeTask('sub', 4, 6)),
  ]);
  console.log(result);
})();
```

**Comment by ggoodman** [comment]
@jasnell I'm sorry for not circling back to this. I had intended to take a stab at a PR.

I'm surprised that it would be that easy (or even work at all). How is the messaging protocol between the `index.js` and `worker.js` files being pulled in? I thought that there was some required wrapper code to accept `"message"` events, dispatch those and then notify the pool manager re completion.

That was the part of the equation that I found difficult to set up in a bundling scenario. The messaging protocol from within the workers seemed to get added implicitly and that wrapping happened using filesystem access relative to the expected location of Piscina's entrypoint. If Piscina is bundled, then paths relative to the entrypoint are no longer meaningful.

**Comment by jasnell** [comment]
Piscina handles all of the messaging back and forth. It spawns the worker thread which is responsible for either importing or requiring the worker code. Piscina creates an internal message object that is passed via postMessage into the worker. The piscina logic in the worker receives the message and calls the exported worker function with the message data. So for this example, the helper simply encodes the operation name in the message payload itself and the helper dispatcher extracts that and forwards it on to the appropriate handler. Nice and simple.

As for any bundler, so long as the bundler is doing the right thing with regards to import and require, it *should* just work. But, I haven't tried it as I don't ever use a bundler on the server side.

**Comment by ggoodman** [comment]
The line that appears to be tripping up esbuild is this: https://github.com/piscinajs/piscina/blob/6de046036609f84a001d5e419fb29429eb323a12/src/index.ts#L524. That type of reference isn't able to convey to a bundler that there is a module dependency. To a bundler, it's just a string with no semantic meaning.

Having read through the code, I felt like the API I had in mind might be very at odds with the current design. Notably how the pool consumer can choose different worker scripts for different tasks.

Maybe I could simply wrap Piscina. However, to do that cleanly would mean putting the worker in the export map as well. What do you think about that?

**Comment by jasnell** [comment]
If you can make it work, then go for it

**Comment by jasnell** [comment]
Just landed the ability to expose multiple worker functions per file. I believe it at least gets us closer to what is being requested here...

example worker:

```js
function add({ a, b }) { return a + b; }
function multiply({ a, b }) { return a * b; }

module.exports = { add, multiply };
```

**Comment by merceyz** [comment]
In https://github.com/yarnpkg/berry/pull/2683 I added a really simple, probably inefficient, worker pool to Yarn. It's used to run tgz to zip conversion in parallel to speed up the fetch step of installs and since Yarn is a single JS file I can't point the Worker class at a file so I had to use the `eval` option and give it the source code as a string.

If Piscina supported telling it how to construct the Worker or rather give it the source code of the worker, that would allow us to switch over to using it instead of the naïve worker pool I created.

The first thing that comes to mind based on https://github.com/piscinajs/piscina/issues/106#issuecomment-782517224 would be something like this:
```ts
// Worker.ts
import 'piscina/worker'
import { registerTask } from 'piscina/worker_helper';

registerTask('convertToZip', async (tgzData: Buffer): Promise<string> => {
	// Run convert and write to disk
	return '/tmp/converted.zip';
});

// index.ts
(async () => {
	const piscina = new Piscina({
		source: 'Bundled Worker.ts source as a string',
	});

	const zipFile = await piscina.runTask('convertToZip', Buffer.from('data'));
	console.log(zipFile);
})();
```
